client/order/epochqueue.go

```diff
@@ -0,0 +1,190 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package order
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"sync"
+
+	"decred.org/dcrdex/dex/msgjson"
+	"decred.org/dcrdex/dex/order"
+	"github.com/decred/dcrd/crypto/blake256"
+)
+
+// EpochQueue represents a client epoch queue.
+type EpochQueue struct {
+	orders             map[order.OrderID]*msgjson.EpochOrderNote
+	ordersMtx          sync.Mutex
+	seedIndex          []order.OrderID
+	seedIndexMtx       sync.Mutex
+	commitmentIndex    []order.OrderID
+	commitmentIndexMtx sync.Mutex
+}
+
+// NewEpochQueue creates a client epoch queue.
+func NewEpochQueue() *EpochQueue {
+	return &EpochQueue{
+		orders:          make(map[order.OrderID]*msgjson.EpochOrderNote),
+		seedIndex:       make([]order.OrderID, 0),
+		commitmentIndex: make([]order.OrderID, 0),
+	}
+}
+
+// Reset clears the epoch queue. This should be called when a new epoch begins.
+func (eq *EpochQueue) Reset() {
+	eq.ordersMtx.Lock()
+	eq.orders = make(map[order.OrderID]*msgjson.EpochOrderNote)
+	eq.ordersMtx.Unlock()
+
+	eq.seedIndexMtx.Lock()
+	eq.seedIndex = make([]order.OrderID, 0)
+	eq.seedIndexMtx.Unlock()
+
+	eq.commitmentIndexMtx.Lock()
+	eq.commitmentIndex = make([]order.OrderID, 0)
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Queue appends the provided note to the epoch queue.
+func (eq *EpochQueue) Queue(note *msgjson.EpochOrderNote) {
+	var oid order.OrderID
+	copy(oid[:], note.OrderID[:])
+
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	eq.orders[oid] = note
+
+	// Update the seed sort order per order ids.
+	eq.seedIndexMtx.Lock()
+	si := sort.Search(len(eq.seedIndex), func(i int) bool {
+		order := eq.orders[eq.seedIndex[i]]
+		return bytes.Compare(order.OrderID[:], oid[:]) > 0
```

Comment:
As we discussed in Matrix chat, it looks like `eq.orders` is not needed here.  i.e. `eq.orders[eq.seedIndex[i]].OrderID == eq.seedIndex[i]`, which simplifies the insertion location search.